File: sdk/crewlayer/_actions.py
"""Actions resource clients — sync and async."""
from __future__ import annotations

from typing import Any

from crewlayer._http import AsyncTransport, SyncTransport
from crewlayer._types import ActionPage, ActionRecord, ActionStats


class ActionsClient:
    """Synchronous action log operations."""

    def __init__(self, http: SyncTransport) -> None:
        self._http = http

    def log(
        self,
        agent_id: str,
        tool_name: str,
        input_params: dict[str, Any],
        output_result: dict[str, Any],
        status: str,
        *,
        session_id: str | None = None,
        duration_ms: int | None = None,
        error_msg: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> ActionRecord:
        """Record an immutable action entry for the agent."""
        data = self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/actions",
            json={
                "tool_name": tool_name,
                "input_params": input_params,
                "output_result": output_result,
                "status": status,
                "session_id": session_id,
                "duration_ms": duration_ms,
                "error_msg": error_msg,
                "metadata": metadata or {},
            },
        )
        return ActionRecord._from(data)

    def get(self, agent_id: str, action_id: str) -> ActionRecord:
        """Retrieve a single action by ID."""
        data = self._http.request("GET", f"/v1/agents/{agent_id}/actions/{action_id}")
        return ActionRecord._from(data)

    def list(
        self,
        agent_id: str,
        *,
        tool: str | None = None,
        status: str | None = None,
        since: str | None = None,
        until: str | None = None,
        limit: int = 50,
        cursor: str | None = None,
    ) -> ActionPage:
        """List actions with optional filters. Paginated via cursor."""
        params: dict[str, Any] = {"limit": limit}
        if tool is not None:
            params["tool"] = tool
        if status is not None:
            params["status"] = status
        if since is not None:
            params["since"] = since
        if until is not None:
            params["until"] = until
        if cursor is not None:
            params["cursor"] = cursor
        data = self._http.request("GET", f"/v1/agents/{agent_id}/actions", params=params)
        return ActionPage._from(data)

    def stats(self, agent_id: str) -> ActionStats:
        """Aggregate statistics: totals, error rate, average duration, per-tool breakdown."""
        data = self._http.request("GET", f"/v1/agents/{agent_id}/actions/stats")
        return ActionStats._from(data)


class AsyncActionsClient:
    """Asynchronous action log operations."""

    def __init__(self, http: AsyncTransport) -> None:
        self._http = http

    async def log(
        self,
        agent_id: str,
        tool_name: str,
        input_params: dict[str, Any],
        output_result: dict[str, Any],
        status: str,
        *,
        session_id: str | None = None,
        duration_ms: int | None = None,
        error_msg: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> ActionRecord:
        """Record an immutable action entry for the agent."""
        data = await self._http.request(
            "POST",
            f"/v1/agents/{agent_id}/actions",
            json={
                "tool_name": tool_name,
                "input_params": input_params,
                "output_result": output_result,
                "status": status,
                "session_id": session_id,
                "duration_ms": duration_ms,
                "error_msg": error_msg,
                "metadata": metadata or {},
            },
        )
        return ActionRecord._from(data)

    async def get(self, agent_id: str, action_id: str) -> ActionRecord:
        """Retrieve a single action by ID."""
        data = await self._http.request("GET", f"/v1/agents/{agent_id}/actions/{action_id}")
        return ActionRecord._from(data)

    async def list(
        self,
        agent_id: str,
        *,
        tool: str | None = None,
        status: str | None = None,
        since: str | None = None,
        until: str | None = None,
        limit: int = 50,
        cursor: str | None = None,
    ) -> ActionPage:
        """List actions with optional filters. Paginated via cursor."""
        params: dict[str, Any] = {"limit": limit}
        if tool is not None:
            params["tool"] = tool
        if status is not None:
            params["status"] = status
        if since is not None:
            params["since"] = since
        if until is not None:
            params["until"] = until
        if cursor is not None:
            params["cursor"] = cursor
        data = await self._http.request("GET", f"/v1/agents/{agent_id}/actions", params=params)
        return ActionPage._from(data)

    async def stats(self, agent_id: str) -> ActionStats:
        """Aggregate statistics: totals, error rate, average duration, per-tool breakdown."""
        data = await self._http.request("GET", f"/v1/agents/{agent_id}/actions/stats")
        return ActionStats._from(data)

File: sdk-typescript/src/resources/actions.ts
import type { HttpClient } from "../http.js";
import type { ActionRecord, ActionPage, ActionStats, ActionStatus } from "../types.js";

export interface LogActionParams {
  agentId: string;
  toolName: string;
  inputParams?: Record<string, unknown>;
  outputResult?: Record<string, unknown>;
  status?: ActionStatus;
  sessionId?: string;
  durationMs?: number;
  errorMsg?: string;
  metadata?: Record<string, unknown>;
}

export interface ListActionsParams {
  agentId: string;
  status?: ActionStatus;
  toolName?: string;
  limit?: number;
  cursor?: string;
  sessionId?: string;
}

export class ActionsResource {
  constructor(private readonly http: HttpClient) {}

  /** Log a new action for an agent. */
  async log(params: LogActionParams): Promise<ActionRecord> {
    const { agentId, ...rest } = params;
    return this.http.post<ActionRecord>(`/v1/agents/${agentId}/actions`, {
      tool_name: rest.toolName,
      input_params: rest.inputParams ?? {},
      output_result: rest.outputResult ?? {},
      status: rest.status ?? "success",
      session_id: rest.sessionId,
      duration_ms: rest.durationMs,
      error_msg: rest.errorMsg,
      metadata: rest.metadata ?? {},
    });
  }

  /** Retrieve a specific action by ID. */
  async get(agentId: string, actionId: string): Promise<ActionRecord> {
    return this.http.get<ActionRecord>(`/v1/agents/${agentId}/actions/${actionId}`);
  }

  /** List actions for an agent with optional filtering and cursor pagination. */
  async list(params: ListActionsParams): Promise<ActionPage> {
    const { agentId, status, toolName, limit, cursor, sessionId } = params;
    return this.http.get<ActionPage>(`/v1/agents/${agentId}/actions`, {
      params: {
        status,
        tool_name: toolName,
        limit,
        cursor,
        session_id: sessionId,
      },
    });
  }

  /** Get aggregated action statistics grouped by tool name. */
  async stats(agentId: string): Promise<ActionStats> {
    return this.http.get<ActionStats>(`/v1/agents/${agentId}/actions/stats`);
  }
}
